src/bp/ui-shared/src/ContextMenu/index.tsx

```diff
@@ -0,0 +1,161 @@
+        // when offset changes, to force recomputing position.
+        const key = this.state.offset == null ? "" : `${this.state.offset.left}x${this.state.offset.top}`;
+
+        // wrap the popover in a positioned div to make sure it is properly
+        // offset on the screen.
+        return (
+            <div className={Classes.CONTEXT_MENU_POPOVER_TARGET} style={this.state.offset}>
+                <Popover
+                    {...this.props}
+                    backdropProps={{ onContextMenu: this.handleBackdropContextMenu }}
+                    content={content}
+                    enforceFocus={false}
+                    key={key}
+                    hasBackdrop={true}
+                    isOpen={this.state.isOpen}
+                    minimal={true}
+                    modifiers={POPPER_MODIFIERS}
+                    onInteraction={this.handlePopoverInteraction}
+                    position={Position.RIGHT_TOP}
+                    popoverClassName={popoverClassName}
+                    target={<div />}
+                    transitionDuration={TRANSITION_DURATION}
+                />
+            </div>
+        );
+    }
+
+    public show(menu: JSX.Element, offset: IOffset, onClose?: () => void, isDarkTheme?: boolean) {
+        this.setState({ isOpen: true, menu, offset, onClose, isDarkTheme });
+    }
+
+    public hide() {
+        safeInvoke(this.state.onClose);
+        this.setState({ isOpen: false, onClose: undefined });
+    }
+
+    private cancelContextMenu = (e: React.SyntheticEvent<HTMLDivElement>) => e.preventDefault();
+
+    private handleBackdropContextMenu = (e: React.MouseEvent<HTMLDivElement>) => {
+        // React function to remove from the event pool, useful when using a event within a callback
+        e.persist();
+        e.preventDefault();
+        // wait for backdrop to disappear so we can find the "real" element at event coordinates.
+        // timeout duration is equivalent to transition duration so we know it's animated out.
+        this.setTimeout(() => {
+            // retrigger context menu event at the element beneath the backdrop.
+            // if it has a `contextmenu` event handler then it'll be invoked.
+            // if it doesn't, no native menu will show (at least on OSX) :(
+            const newTarget = document.elementFromPoint(e.clientX, e.clientY);
+            newTarget.dispatchEvent(new MouseEvent("contextmenu", e));
+        }, TRANSITION_DURATION);
+    };
+
+    private handlePopoverInteraction = (nextOpenState: boolean) => {
+        if (!nextOpenState) {
+            // delay the actual hiding till the event queue clears
+            // to avoid flicker of opening twice
+            requestAnimationFrame(() => this.hide());
+        }
+    };
+}
+
+let contextMenuElement: HTMLElement;
+let contextMenu: ContextMenu;
+*/
+import React, { FC, Fragment, SyntheticEvent, useEffect } from 'react'
+import ReactDOM from 'react-dom'
+
+import Overlay from '../Overlay'
```

Comment:
Sorry I hink I missed that, what's overlay ? Is it a foreground that darkens the whole page ?

Reply:
It's just to handle the click outside of the element in the foreground (ex.: context menu). Instead of adding a listener and checking that the target isn't part of the element in the foreground (context menu) we just have two elements, one covering the entire document and another one on top of it, the background has a simple onClick to handle clicks that aren't on the element on the foreground. For now, it's transparent, but a darkened shade can be added in the future if needed.